File: go/hackerrank/array/sparse_array.go
package array

func matchingStrings(strings []string, queries []string) []int32 {
	res := make([]int32, len(queries))
	dict := make(map[string]int32, 0)

	for _, value := range strings {
		if _, ok := dict[value]; ok {
			dict[value] = dict[value] + 1
		} else {
			dict[value] = 1
		}
	}
	for index, q := range queries {
		if count, ok := dict[q]; ok {
			res[index] = count
		} else {
			res[index] = 0
		}
	}

	return res
}

File: python/hackerrank/array/sparse_array.py
def matchingStrings(strings, queries):
    my_dict = {}
    for item in strings:
        if item in my_dict:
            my_dict[item] = my_dict[item] + 1
        else:
            my_dict[item] = 1
    res = []
    for item in queries:
        if item in my_dict:
            res.append(my_dict[item])
        else:
            res.append(0)
    return res

strings = ["aba", "baba", "aba", "xzxb"]
queries = ["aba", "xzxb", "ab"]

print(matchingStrings(strings, queries))